src/dynarec/dynarec_arm_d9.c

```diff
@@ -449,32 +514,63 @@ uintptr_t dynarecD9(dynarec_arm_t* dyn, uintptr_t addr, uintptr_t ip, int ninst,
             INST_NAME("FYL2XP1");
             v1 = x87_get_st(dyn, ninst, x1, x2, 0, NEON_CACHE_ST_D);
             v2 = x87_get_st(dyn, ninst, x1, x2, 1, NEON_CACHE_ST_D);
-            VMOV_i_64(0, 0b01110000);   // D0 = 1.0
-            VADD_F64(0, 0, v1);    // prepare call to log2
-            CALL_1D(log2, 0);
-            VMUL_F64(v2, v2, 0);    //ST(1).d = log2(ST0.d + 1.0)*ST(1).d;
+            if(!box86_dynarec_fastround) {
+                LDRH_IMM8(x1, xEmu, offsetof(x86emu_t, cw));    // hopefully cw is not too far for an imm8
```

Comment:
I don't understand this part. Why extracting the cw flags for then hacking it based on sign of the number. Are you sure there isn't a simpler way and just applying a swizzled cw not enough?

Reply:
When ST1 < 0: larger log1p(ST0) leads to smaller ST1*log1p(ST0), and smaller log1p(ST0) leads to larger ST1*log1p(ST0).
So when the rounding mode of x86emu is upward, downward when computing log1p(ST0) and upward when computing the multiplication is needed to ensure result of ST1*log1p(ST0) > its precise value (in math).

Reply:
So when before log1p is called, the rounding mode is set to the opposite one:
- toward 0 -> toward 0
- toward up -> toward down
- toward down -> toward up
- toward nearest->toward nearest

Then after log1p but before the multiplication, the rounding mode is set as x86emu (`x87_setround`).

Reply:
but is all this really needed? the aim is to behave like x87, ok, but this lead to so many operations. Is all this really done by the x87 natively?

Reply:
> but is all this really needed?

As for as for Hisoutensoku (Touhou 12.3), not really needed.

> Is all this really done by the x87 natively?

X87 gives rounded result. And this "two-step" rounding is trying to emulate its result, but still not precise as x87, because it computes in two steps and every step reduces accuracy.

It is OK for me to give up these two-step rounding. They are just incidentally written when I was writing rounding code for other instructions.

Reply:
Well, if there is no current case wee it's usefull, I would prefer to keep it as simple as possible yes.

Reply:
What do you think of it? Should we use a way which is simpler and faster but also a bit less precise?
For example, we can apply the rounding mode only on multiplication, then we don't even need to use stack to store the origin fpscr value; we can also simply use the default rounding mode without setting it.

Reply:
The simpler the better. Precision is probably not that important here. Not sure x87 is actually more precise.

Reply:
> we can also simply use the default rounding mode without setting it.

Done.